### PR TITLE
Handle invalid routes better

### PR DESCRIPTION
**Pull request by lpichler**
as done here before https://github.com/RedHatInsights/topological_inventory-api/pull/188 

cc @slemrmartin

### CODE REVIEW

File: app/controllers/api.rb [hunk 1]
```diff
@@ -1,2 +1,10 @@
 module Api
+  class RootController < ApplicationController
+    skip_before_action :validate_primary_collection_id
+
```
Comment (slemrmartin) [inline app/controllers/api.rb]: @lpichler POST requests raised some error, please try

Comment (lpichler) [reply app/controllers/api.rb]: fixed, thanks 👍

**Comment by lpichler** [comment]
@abellotti we want to [return the proper error message for invalid URLs](https://github.com/RedHatInsights/topological_inventory-api/pull/188) as you did before.

and there is issue with HEAD request:

1.  when you do HEAD request and there is route in `routes.rb`
 ```ruby
match "*path", :to => "api/root#invalid_url_error", :via =>ActionDispatch::Routing::HTTP_METHODS
```
=> **request will end in root controller**


2.  when you do HEAD request and there is **not present** route from point 1.  in `routes.rb`
=> **request will end in related resource controller** (this is expected)

So in second commit, I added code to handle HEAD request in root controller and check if we have such route in GET and it will fails in not. 

What do you think about it ?

thanks!

**Comment by abellotti** [comment]
Finally got around to testing this a bit, sorry about the delay @lpichler. for an active record not found, while the 404 is properly returned, I'm getting the "Invalid URL /api/v2.0/sources/123 specified." which is  misleading. it's a correct URL, error message should just be "Not found" or "Record not found". Thanks

File: app/controllers/api.rb [hunk 2]
```diff
@@ -1,2 +1,18 @@
...
+    skip_before_action :validate_primary_collection_id
+    skip_before_action :validate_request
+
+    def head_request_exists_in_get_routes?
+      routes = Rails.application.routes.recognize_path(request.path, :method => :get)
+      routes.extract!(:controller, :action) != {:controller => "api/root", :action => "invalid_url_error"}
+    end
+
+    def invalid_url_error
+      return head :ok if request.head? && head_request_exists_in_get_routes?
```
Comment (bdunne) [inline app/controllers/api.rb]: Should this ever render `:ok`?  We only get here is the url is invalid, right?

Comment (syncrou) [reply app/controllers/api.rb:12]: @lpichler would you be able to provide any followup to @bdunne?

Comment (lpichler) [reply app/controllers/api.rb]: I removed logic for `HEAD` request because it need to solved differently - maybe by adding head routes to `routes.rb`

Comment (lpichler) [reply app/controllers/api.rb]: I removed logic for HEAD and we need to investigate whether there is better way how to handle HEAD request.

**Comment by lpichler** [comment]
> Finally got around to testing this a bit, sorry about the delay @lpichler. for an active record not found, while the 404 is properly returned, I'm getting the "Invalid URL /api/v2.0/sources/123 specified." which is misleading. it's a correct URL, error message should just be "Not found" or "Record not found". Thanks

@abellotti 
we are using in specs format

`/api/v1.0/sources/123`

but it looks like that correct url is:

`/api/sources/v1.0/sources` and `/api/sources/v1.0/` is prefix.

then `/api/v2.0/sources/123` is not correct and this `/api/sources/v2.0/sources/123` returns not found as expected.

@slemrmartin which format is correct ?

**Comment by lpichler** [comment]
Ready to re-review 👍 

thanks

**Comment by slemrmartin** [comment]
@lpichler is this PR still active?

**Pull request closed without merging**